### PR TITLE
Nsg e2e

### PR DESCRIPTION
**Pull request by bryanro92**
### Which issue this PR addresses:

Fixes https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/10542504/

### What this PR does / why we need it:
Adds a test that creates a new NSG, associates it with worker subnet, and then waits for nsg controller to reconcile and replace with correct ARO nsg

### Test plan for issue:
e2e passes

### Is there any documentation that needs to be updated for this PR?
nope

### CODE REVIEW

File: test/e2e/operator.go [hunk 1]
```diff
@@ -272,3 +277,99 @@ var _ = Describe("ARO Operator - Conditions", func() {
...
+		resourceGroup string
+		vnetName      string
+		subnetName    string
+		location      string
+	}
+	BeforeEach(func() {
+		enableReconcileNSG := func() {
+			instance, err := clients.AROClusters.AroV1alpha1().Clusters().Get(context.Background(), arov1alpha1.SingletonClusterName, metav1.GetOptions{})
+			Expect(err).NotTo(HaveOccurred())
+			instance.Spec.Features.ReconcileNSGs = true
```
Comment (mjudeikis) [inline test/e2e/operator.go:295]: Check if we need to enable. Once this becomes default we will not. So avoid future call :)

File: test/e2e/operator.go [hunk 2]
```diff
@@ -272,3 +277,99 @@ var _ = Describe("ARO Operator - Conditions", func() {
...
+	var correctNSG *string
+	var dummynsg mgmtnetwork.SecurityGroup
+	type subnetDescriptor struct {
+		resourceGroup string
+		vnetName      string
+		subnetName    string
+		location      string
+	}
+	BeforeEach(func() {
+		enableReconcileNSG := func() {
```
Comment (mjudeikis) [inline test/e2e/operator.go:292]: Overall bit strange pattern to define functions like this here and after this just call them once. 
Either just do all this in the before or define functions outside and call them. confusing to read.

**Review by mjudeikis — COMMENTED** [review]
nits

File: test/e2e/operator.go (hunk 1 above)
Comment (bryanro92) [reply test/e2e/operator.go]: good call, changed to only update instance when `!ReconcileNSGs`

File: test/e2e/operator.go (hunk 2 above)
Comment (bryanro92) [reply test/e2e/operator.go]: moved the funcs above BeforeEach. I added in the funcs to try and break it down so it was more clear the different steps/tasks I was doing.

File: test/e2e/operator.go [hunk 3]
```diff
@@ -272,3 +277,101 @@ var _ = Describe("ARO Operator - Conditions", func() {
...
+	var correctNSG *string
+	var dummynsg mgmtnetwork.SecurityGroup
+	type subnetDescriptor struct {
+		resourceGroup string
+		vnetName      string
+		subnetName    string
+		location      string
+	}
+	enableReconcileNSG := func() {
+		instance, err := clients.AROClusters.AroV1alpha1().Clusters().Get(context.Background(), arov1alpha1.SingletonClusterName, metav1.GetOptions{})
```
Comment (cadenmarchese) [inline test/e2e/operator.go:292]: Someone on the team gave me this feedback on an e2e (maybe @troy0820), so thought I'd pass it along.

Rather than use `context.Background()` in each Get/List/Update, you can define it in a `ctx := context.Background() ` var and use it like you would outside of a test:
```suggestion
		instance, err := clients.AROClusters.AroV1alpha1().Clusters().Get(ctx, arov1alpha1.SingletonClusterName, metav1.GetOptions{})
```
It's functionally the same since it's just an empty context, but avoids having to define a new empty context on each Get/List/Update.

File: test/e2e/operator.go [hunk 4]
```diff
@@ -272,3 +277,101 @@ var _ = Describe("ARO Operator - Conditions", func() {
...
+		}
+		return &spec.ResourceGroup, &subnetDescriptor{
+			resourceGroup: spec.NetworkResourceGroup,
+			vnetName:      spec.Vnet,
+			subnetName:    spec.Subnet,
+			location:      spec.Location,
+		}, nil
+	}
+	gatherNetworkInfo := func() {
+		machines, err := clients.MachineAPI.MachineV1beta1().Machines("openshift-machine-api").List(context.Background(), metav1.ListOptions{LabelSelector: "machine.openshift.io/cluster-api-machine-role=worker"})
```
Comment (cadenmarchese) [inline test/e2e/operator.go:314]:
```suggestion
		machines, err := clients.MachineAPI.MachineV1beta1().Machines("openshift-machine-api").List(ctx, metav1.ListOptions{LabelSelector: "machine.openshift.io/cluster-api-machine-role=worker"})
```

File: test/e2e/operator.go [hunk 5]
```diff
@@ -272,3 +277,101 @@ var _ = Describe("ARO Operator - Conditions", func() {
...
+		}
+	}
+	createDummyNSG := func() {
+		dummynsg = mgmtnetwork.SecurityGroup{
+			Location:                      to.StringPtr(location),
+			Name:                          to.StringPtr("dummynsg"),
+			Type:                          to.StringPtr("Microsoft.Network/networkSecurityGroups"),
+			SecurityGroupPropertiesFormat: &mgmtnetwork.SecurityGroupPropertiesFormat{},
+		}
+		err := clients.NetworkSecurityGroups.CreateOrUpdateAndWait(context.Background(), resourceGroup, "dummynsg", dummynsg)
```
Comment (cadenmarchese) [inline test/e2e/operator.go:338]:
```suggestion
		err := clients.NetworkSecurityGroups.CreateOrUpdateAndWait(ctx, resourceGroup, "dummynsg", dummynsg)
```

File: test/e2e/operator.go [hunk 6]
```diff
@@ -272,3 +277,101 @@ var _ = Describe("ARO Operator - Conditions", func() {
...
+		Expect(err).NotTo(HaveOccurred())
+		dummynsgObject, err := clients.NetworkSecurityGroups.Get(context.Background(), resourceGroup, "dummynsg", "")
```
Comment (cadenmarchese) [inline test/e2e/operator.go:340]:
```suggestion
		dummynsgObject, err := clients.NetworkSecurityGroups.Get(ctx, resourceGroup, "dummynsg", "")
```

File: test/e2e/operator.go [hunk 7]
```diff
@@ -272,3 +277,101 @@ var _ = Describe("ARO Operator - Conditions", func() {
...
+		Expect(err).NotTo(HaveOccurred())
+		dummynsg.ID = dummynsgObject.ID
+	}
+	BeforeEach(func() {
+		enableReconcileNSG()
+		gatherNetworkInfo()
+		createDummyNSG()
+	})
+	AfterEach(func() {
+		err := clients.NetworkSecurityGroups.DeleteAndWait(context.Background(), resourceGroup, "dummynsg")
```
Comment (cadenmarchese) [inline test/e2e/operator.go:350]:
```suggestion
		err := clients.NetworkSecurityGroups.DeleteAndWait(ctx, resourceGroup, "dummynsg")
```

File: test/e2e/operator.go [hunk 8]
```diff
@@ -272,3 +277,101 @@ var _ = Describe("ARO Operator - Conditions", func() {
...
+		if err != nil {
+			log.Warn(err)
+		}
+	})
+	Specify("If worker nsg is changed or removed then controller should reconcile", func() {
+		validateNSG := func() (bool, error) {
+			s, err := clients.Subnet.Get(context.Background(), resourceGroup, vnet, subnet, "")
```
Comment (cadenmarchese) [inline test/e2e/operator.go:357]:
```suggestion
			s, err := clients.Subnet.Get(ctx, resourceGroup, vnet, subnet, "")
```

File: test/e2e/operator.go [hunk 9]
```diff
@@ -272,3 +277,101 @@ var _ = Describe("ARO Operator - Conditions", func() {
...
+				return false, err
+			}
+			if *s.NetworkSecurityGroup.ID == *correctNSG {
+				log.Info("nsg matched expected value")
+				return true, nil
+			}
+			log.Info("nsg did not match expected value")
+			return false, nil
+		}
+		subnetObject, err := clients.Subnet.Get(context.Background(), resourceGroup, vnet, subnet, "")
```
Comment (cadenmarchese) [inline test/e2e/operator.go:368]:
```suggestion
		subnetObject, err := clients.Subnet.Get(ctx, resourceGroup, vnet, subnet, "")
```

File: test/e2e/operator.go [hunk 10]
```diff
@@ -272,3 +277,101 @@ var _ = Describe("ARO Operator - Conditions", func() {
...
+		Expect(err).NotTo(HaveOccurred())
+		correctNSG = subnetObject.NetworkSecurityGroup.ID
+		subnetObject.NetworkSecurityGroup = &dummynsg
+		err = clients.Subnet.CreateOrUpdateAndWait(context.Background(), resourceGroup, vnet, subnet, subnetObject)
```
Comment (cadenmarchese) [inline test/e2e/operator.go:372]:
```suggestion
		err = clients.Subnet.CreateOrUpdateAndWait(ctx, resourceGroup, vnet, subnet, subnetObject)
```

File: test/e2e/operator.go [hunk 11]
```diff
@@ -132,7 +137,7 @@ var _ = Describe("ARO Operator - Internet checking", func() {
 })
 
 var _ = Describe("ARO Operator - Geneva Logging", func() {
-	Specify("genevalogging must be repaired if deployment deleted", func() {
+	Specify("genevalogging must be repaired if daemonset deleted", func() {
 		mdsdReady := func() (bool, error) {
```
Comment (cadenmarchese) [inline test/e2e/operator.go:141]:
```suggestion
		ctx := context.Background()
		
		mdsdReady := func() (bool, error) {
```

**Review by cadenmarchese — APPROVED** [review]

File: pkg/util/azureclient/mgmt/network/securitygroups_addons.go [hunk 12]
```diff
@@ -24,6 +25,14 @@ func (c *securityGroupsClient) CreateOrUpdateAndWait(ctx context.Context, resour
 	return future.WaitForCompletionRef(ctx, c.Client)
 }
 
+func (c *securityGroupsClient) DeleteAndWait(ctx context.Context, resourceGroupName string, networkSecurityGroupName string) (err error) {
+	future, err := c.Delete(ctx, resourceGroupName, networkSecurityGroupName)
```
Comment (petrkotas) [inline pkg/util/azureclient/mgmt/network/securitygroups_addons.go:29]: In the light of https://github.com/Azure/ARO-RP/pull/1679 would you please reference the original client on this line? 
```suggestion
	future, err := c.SecurityGroupsClient.Delete(ctx, resourceGroupName, networkSecurityGroupName)
```

File: test/e2e/operator.go [hunk 13]
```diff
@@ -272,3 +277,102 @@ var _ = Describe("ARO Operator - Conditions", func() {
...
+		Expect(err).NotTo(HaveOccurred())
+		subnetMap := make(map[subnetDescriptor]bool)
+		for _, machine := range machines.Items {
+			_, subnetDesc, err := getDescriptorFromProviderSpec(machine.Spec.ProviderSpec.Value)
+			if err != nil {
+				log.Fatal(err)
+			}
+			subnetMap[*subnetDesc] = true
+		}
+		for ws := range subnetMap {
```
Comment (petrkotas) [inline test/e2e/operator.go:325]: Q: Is the length of `subnetMap` always equal to 1?

File: test/e2e/operator.go [hunk 14]
```diff
@@ -272,3 +277,102 @@ var _ = Describe("ARO Operator - Conditions", func() {
...
+		validateNSG := func() (bool, error) {
+			s, err := clients.Subnet.Get(ctx, resourceGroup, vnet, subnet, "")
+			if err != nil {
+				return false, err
+			}
+			if *s.NetworkSecurityGroup.ID == *correctNSG {
+				log.Info("nsg matched expected value")
+				return true, nil
+			}
+			log.Info("nsg did not match expected value")
```
Comment (petrkotas) [inline test/e2e/operator.go:366]: Nit: If the flow reach this end, what value would you think logging with `error` level and `*s.NetworkSecurityGroup.ID` information will bring? How much will it simplify debug with such information at hand?

**Review by petrkotas — CHANGES_REQUESTED** [review]
Thank you @bryanro92 for this PR. This is really nice work. My apologies for nitpicking. Would you please look into my 3 comments?

File: test/e2e/operator.go (hunk 14 above)
Comment (bryanro92) [reply test/e2e/operator.go:366]: I went with `info` at first because it is expected reconciler may need grace period before reconciling, but upon reflection of your comment I agree that any change of the network security group from expected group should throw an error. 

The `*s.NetWorkSecurityGroup.ID` and `*correctNSG` would definitely provide visibility into "why" it did not match expected value. I will update that.

File: test/e2e/operator.go [hunk 15]
```diff
@@ -272,3 +277,102 @@ var _ = Describe("ARO Operator - Conditions", func() {
...
+	}
+	enableReconcileNSG := func() {
+		instance, err := clients.AROClusters.AroV1alpha1().Clusters().Get(ctx, arov1alpha1.SingletonClusterName, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		if !instance.Spec.Features.ReconcileNSGs {
+			instance.Spec.Features.ReconcileNSGs = true
+			_, err = clients.AROClusters.AroV1alpha1().Clusters().Update(ctx, instance, metav1.UpdateOptions{})
+			Expect(err).NotTo(HaveOccurred())
+		}
+	}
```
Comment (m1kola) [inline test/e2e/operator.go:300]: Hm. Why do we have it disabled by default? Is it intentional?

cc @Makdaam

File: test/e2e/operator.go [hunk 16]
```diff
@@ -272,3 +277,102 @@ var _ = Describe("ARO Operator - Conditions", func() {
...
+			}
+			subnetMap[*subnetDesc] = true
+		}
+		for ws := range subnetMap {
+			location = ws.location
+			resourceGroup = ws.resourceGroup
+			subnet = ws.subnetName
+			vnet = ws.vnetName
+		}
+	}
```
Comment (m1kola) [inline test/e2e/operator.go:331]: Do we actually need all this? My understanding of what this code is doing:

1. It lists all worker machine sets
2. It gathers unique subnets (`subnetMap`)
3. Then it take only one subnet from `subnetMap` and the further tests work with it. 

Why do we gather unique subnets if we need only one?

File: test/e2e/operator.go [hunk 17]
```diff
@@ -272,3 +277,102 @@ var _ = Describe("ARO Operator - Conditions", func() {
...
+		}
+		return &spec.ResourceGroup, &subnetDescriptor{
+			resourceGroup: spec.NetworkResourceGroup,
+			vnetName:      spec.Vnet,
+			subnetName:    spec.Subnet,
+			location:      spec.Location,
+		}, nil
+	}
+	gatherNetworkInfo := func() {
+		machines, err := clients.MachineAPI.MachineV1beta1().Machines("openshift-machine-api").List(ctx, metav1.ListOptions{LabelSelector: "machine.openshift.io/cluster-api-machine-role=worker"})
```
Comment (m1kola) [inline test/e2e/operator.go:315]: This reconciler reconciles both worker and master subnets. Is there a benefit of testing both in e2e?

File: test/e2e/operator.go (hunk 16 above)
Comment (m1kola) [reply test/e2e/operator.go:331]: I think we don't need to read machinesets here at all. Can we do something like this instead?

1. Read cluster object to get vnet id.
2. Get vnet object which contains all subnets
3. Modify one of the subnets.

See [`[Admin API] List Azure resources action`](https://github.com/Azure/ARO-RP/blob/e85e4a07adef15378b350f913800ae90cb2af018/test/e2e/adminapi_resources.go) for a refrence.

I have a feeling that it will significantly simplify the test.

**Review by m1kola — CHANGES_REQUESTED** [review]
This test is too complex. At this point we need a test for this test :) Let's try and simplify.

File: test/e2e/operator.go (hunk 15 above)
Comment (m1kola) [reply test/e2e/operator.go:300]: Also https://github.com/Azure/ARO-RP/pull/1663 changes the relevant controller and the flag.

Comment (m1kola) [reply test/e2e/operator.go:300]: > Hm. Why do we have it disabled by default? Is it intentional?

Ah, right. I guess this is due to ArchitectureVersion not being respected.

File: test/e2e/operator.go [hunk 18]
```diff
@@ -272,3 +277,102 @@ var _ = Describe("ARO Operator - Conditions", func() {
...
+		gatherNetworkInfo()
+		createDummyNSG()
+	})
+	AfterEach(func() {
+		err := clients.NetworkSecurityGroups.DeleteAndWait(context.Background(), resourceGroup, "dummynsg")
+		if err != nil {
+			log.Warn(err)
+		}
+	})
+	Specify("If worker nsg is changed or removed then controller should reconcile", func() {
```
[the remark this answers is not in the document]
Comment (bryanro92) [reply test/e2e/operator.go:356]: I actually started by testing of just removing NSG, but the reconciler actually does not reconcile if the NSG is _removed_. It only reconciles right now if the NSG is _changed_. I created this item for that bug https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/10554108/

File: test/e2e/operator.go (hunk 17 above)
Comment (bryanro92) [reply test/e2e/operator.go:315]: Yes we should probably test both. I will need to modify the `dummyNSG` properties because when I changed it on the master subnet the tests failed because it could not connect to the cluster anymore.

**Comment by bryanro92** [comment]
definitely just messed up during my rebase, happy Monday. Going to attempt to salvage, but may just close and re-open PR.

File: test/e2e/operator.go (hunk 13 above)
Comment (bryanro92) [reply test/e2e/operator.go]: I have gotten rid of this bit of code so dont use this loop over subnetmap anymore

File: test/e2e/operator.go (hunk 14 above)
Comment (bryanro92) [reply test/e2e/operator.go]: I have added more descriptive and error level logging for when nsg are not equal

File: test/e2e/operator.go (hunk 16 above)
Comment (bryanro92) [reply test/e2e/operator.go]: I have removed this code that relied on logic using in NSG controller and instead now use method you suggested.

File: test/e2e/operator.go (hunk 17 above)
Comment (bryanro92) [reply test/e2e/operator.go]: I separated into testing reconcile on worker subnet and master subnet

**Comment by bryanro92** [comment]
```  
Code="ResourceQuotaExceeded" Message="Resource quota of cores exceeded. Maximum allowed: 200, Current in use: 192, Additional requested: 30."
```
/azp run e2e

File: test/e2e/operator.go (hunk 15 above)
Comment (bryanro92) [reply test/e2e/operator.go:292]: > Also #1663 changes the relevant controller and the flag.

I see feature being changed to `ReconcileSubnets`, I don't mind changing it in this test, but need to understand priority and when each of these plan on being merged so appropriate changes can be made.

File: test/e2e/operator.go [hunk 19]
```diff
@@ -272,3 +276,108 @@ var _ = Describe("ARO Operator - Conditions", func() {
...
 })
+
+var _ = Describe("ARO Operator - Azure NSG Reconciler", func() {
+	var vnetName, location, resourceGroup, workerSubnet, masterSubnet string
+	var dummynsg mgmtnetwork.SecurityGroup
+	ctx := context.Background()
+	enableReconcileNSG := func() {
+		instance, err := clients.AROClusters.AroV1alpha1().Clusters().Get(ctx, arov1alpha1.SingletonClusterName, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		if !instance.Spec.Features.ReconcileNSGs {
```
Comment (mjudeikis) [inline test/e2e/operator.go:287]: Please add TODO and create an gh issue to remove this once it is enabled by default

**Review by mjudeikis — APPROVED** [review]
nits for followup

File: test/e2e/operator.go [hunk 20]
```diff
@@ -272,3 +276,108 @@ var _ = Describe("ARO Operator - Conditions", func() {
...
+		resourceGroup = r.ResourceGroup
+
+		vnet, err := clients.VirtualNetworks.Get(ctx, resourceGroup, r.ResourceName, "")
+		Expect(err).NotTo(HaveOccurred())
+		location = *vnet.Location
+		vnetName = *vnet.Name
+
+		_, masterSubnet, err = subnet.Split(strings.ToLower(*oc.OpenShiftClusterProperties.MasterProfile.SubnetID))
+		Expect(err).NotTo(HaveOccurred())
+		_, workerSubnet, err = subnet.Split(strings.ToLower(*(*oc.OpenShiftClusterProperties.WorkerProfiles)[0].SubnetID))
```
Comment (m1kola) [inline test/e2e/operator.go:310]: I think there is no need to do `strings.ToLower` here. I guess it was a copy-paste from `[Admin API] List Azure resources action` where it is important for deduplication reasons. In our case it is not required, I believe.

File: test/e2e/operator.go [hunk 21]
```diff
@@ -272,3 +276,108 @@ var _ = Describe("ARO Operator - Conditions", func() {
...
+			Location:                      to.StringPtr(location),
+			Name:                          to.StringPtr("dummynsg"),
+			Type:                          to.StringPtr("Microsoft.Network/networkSecurityGroups"),
+			SecurityGroupPropertiesFormat: &mgmtnetwork.SecurityGroupPropertiesFormat{},
+		}
+		err := clients.NetworkSecurityGroups.CreateOrUpdateAndWait(ctx, resourceGroup, "dummynsg", dummynsg)
+		Expect(err).NotTo(HaveOccurred())
+		dummynsgObject, err := clients.NetworkSecurityGroups.Get(ctx, resourceGroup, "dummynsg", "")
+		Expect(err).NotTo(HaveOccurred())
+		dummynsg.ID = dummynsgObject.ID
```
Comment (m1kola) [inline test/e2e/operator.go:324]: Maybe just replace `dummynsg` with a full object?

```suggestion
		dummynsg, err = clients.NetworkSecurityGroups.Get(ctx, resourceGroup, "dummynsg", "")
		Expect(err).NotTo(HaveOccurred())
```

File: test/e2e/operator.go [hunk 22]
```diff
@@ -272,3 +276,108 @@ var _ = Describe("ARO Operator - Conditions", func() {
...
+			if *ms.NetworkSecurityGroup.ID == *correctMasterNSG {
+				log.Info("master subnetnsg matched expected value")
+				return true, nil
+			}
+			log.Errorf("master nsg: %s did not match expected value: %s", *ms.NetworkSecurityGroup.ID, *correctMasterNSG)
+			return false, nil
+		})
+		Expect(err).NotTo(HaveOccurred())
+	})
+})
```
Comment (m1kola) [inline test/e2e/operator.go:383]: These two tests look identical to me. The only difference I is master/worker subnet. Can create a function which accepts params so we can resue it in these two tests?

File: test/e2e/operator.go [hunk 23]
```diff
@@ -272,3 +276,108 @@ var _ = Describe("ARO Operator - Conditions", func() {
...
+		vnetID, _, err := subnet.Split(*oc.OpenShiftClusterProperties.MasterProfile.SubnetID)
+		Expect(err).NotTo(HaveOccurred())
+
+		r, err := azure.ParseResourceID(vnetID)
+		Expect(err).NotTo(HaveOccurred())
+		resourceGroup = r.ResourceGroup
+
+		vnet, err := clients.VirtualNetworks.Get(ctx, resourceGroup, r.ResourceName, "")
+		Expect(err).NotTo(HaveOccurred())
+		location = *vnet.Location
```
Comment (m1kola) [inline test/e2e/operator.go:305]: We can have `location` from `oc.OpenShiftClusterProperties`.

File: test/e2e/operator.go [hunk 24]
```diff
@@ -272,3 +276,108 @@ var _ = Describe("ARO Operator - Conditions", func() {
...
+		vnetName = *vnet.Name
```
Comment (m1kola) [inline test/e2e/operator.go:306]: Why are we doing this? `r.ResourceName` is a `vnetName`. No?

I think we do not need `clients.VirtualNetworks.Get(ctx, resourceGroup, r.ResourceName, "")` call at all.

File: test/e2e/operator.go [hunk 25]
```diff
@@ -272,3 +276,108 @@ var _ = Describe("ARO Operator - Conditions", func() {
...
+		if !instance.Spec.Features.ReconcileNSGs {
+			instance.Spec.Features.ReconcileNSGs = true
+			_, err = clients.AROClusters.AroV1alpha1().Clusters().Update(ctx, instance, metav1.UpdateOptions{})
+			Expect(err).NotTo(HaveOccurred())
+		}
+	}
+	gatherNetworkInfo := func() {
+		oc, err := clients.OpenshiftClustersv20200430.Get(ctx, vnetResourceGroup, clusterName)
+		Expect(err).NotTo(HaveOccurred())
+		vnetID, _, err := subnet.Split(*oc.OpenShiftClusterProperties.MasterProfile.SubnetID)
```
Comment (m1kola) [inline test/e2e/operator.go:296]: Below you are doing `_, masterSubnet, err = subnet.Split(strings.ToLower(*oc.OpenShiftClusterProperties.MasterProfile.SubnetID))`

You can have `masterSubnet` in one go here, I think:

```go
vnetID, masterSubnet, err := subnet.Split(*oc.OpenShiftClusterProperties.MasterProfile.SubnetID)
```

File: test/e2e/operator.go [hunk 26]
```diff
@@ -272,3 +276,108 @@ var _ = Describe("ARO Operator - Conditions", func() {
...
+			ws, err := clients.Subnet.Get(ctx, resourceGroup, vnetName, workerSubnet, "")
+			if err != nil {
+				return false, err
+			}
+			if *ws.NetworkSecurityGroup.ID == *correctWorkerNSG {
+				log.Info("worker subnetnsg matched expected value")
+				return true, nil
+			}
+			log.Errorf("worker nsg: %s did not match expected value: %s", *ws.NetworkSecurityGroup.ID, *correctWorkerNSG)
+			return false, nil
```
Comment (m1kola) [inline test/e2e/operator.go:356]: Should we return an error instead of `nil`. What you put in `log.Errorf` maybe?

**Review by m1kola — CHANGES_REQUESTED** [review]
Looks significantly more readable now. Thanks.

But still we can simplify, I think: we can get rid of some of some api calls and remove duplicating code.

File: test/e2e/operator.go (hunk 19 above)
Comment (bryanro92) [reply test/e2e/operator.go]: https://github.com/Azure/ARO-RP/issues/1735

File: test/e2e/operator.go (hunk 25 above)
Comment (bryanro92) [reply test/e2e/operator.go]: ty fixed

File: test/e2e/operator.go (hunk 24 above)
Comment (bryanro92) [reply test/e2e/operator.go]: good point. fixed.

File: test/e2e/operator.go (hunk 22 above)
Comment (bryanro92) [reply test/e2e/operator.go]: I just decided to loop over steps for each subnet we need to test instead of a func call.

**Comment by bryanro92** [comment]
need to look more into it, test is now failing and see this in aro master operator logs 

```
njb-nsg" func="subnets.(*reconcileManager).ensureSubnetNSG()" file="pkg/operator/controllers/subnets/subnet_nsg.go:49" controller=AzureSubnets
time="2021-09-12T22:32:12Z" level=error msg="Reconciler error network.SubnetsClient#CreateOrUpdate: Failure sending request: StatusCode=0 -- Original Error: Code=\"LinkedAuthorizationFailed\" Message=\"The client '361792a0-ed31-4277-a5c5-
1a4d10e400ea' with object id '361792a0-ed31-4277-a5c5-1a4d10e400ea' has permission to perform action 'Microsoft.Network/virtualNetworks/subnets/write' on scope '/subscriptions/225e02bc-43d0-43d1-a01a-17e584a4ef69/resourceGroups/v4-eastus/
providers/Microsoft.Network/virtualNetworks/dev-vnet/subnets/rbfips2-worker'; however, it does not have permission to perform action 'Microsoft.Network/networkSecurityGroups/join/action' on the linked scope(s) '/providers/Microsoft.Networ
k/networkSecurityGroups/rbfips2-l7njb-nsg' or the linked scope(s) are invalid.\"" func="controller.(*Controller).reconcileHandler()" file="vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:302" AzureSubnets= cont
roller= controller-runtime= manager= name=rbfips2-l7njb-worker-eastus1-mxh7w namespace= reconciler group=aro.openshift.io reconciler kind=Cluster
```

not sure what changed or if just issue with my dev environment

**Comment by petrkotas** [comment]
I have a bad feeling the issues are linked to the OVN networking. I believe older SDN is deploy, I am reruning it again and will see.

File: test/e2e/operator.go [hunk 27]
```diff
@@ -321,3 +325,91 @@ var _ = XDescribe("ARO Operator - MachineSet Controller", func() {
...
+		gatherNetworkInfo()
+		createDummyNSG()
+	})
+	AfterEach(func() {
+		err := clients.NetworkSecurityGroups.DeleteAndWait(context.Background(), resourceGroup, "dummynsg")
+		if err != nil {
+			log.Warn(err)
+		}
+	})
+	Specify("ARO Operator - Azure Subnet Reconciler must reconcile list of subnets when NSG is changed", func() {
```
Comment (m1kola) [inline test/e2e/operator.go:389]: No need to duplicate "ARO Operator - Azure Subnet Reconciler" - it will be printed from the `Describe` block.

```suggestion
	It("must reconcile list of subnets when NSG is changed", func() {
```

File: test/e2e/operator.go [hunk 28]
```diff
@@ -321,3 +325,91 @@ var _ = XDescribe("ARO Operator - MachineSet Controller", func() {
...
+		oc, err := clients.OpenshiftClustersv20200430.Get(ctx, vnetResourceGroup, clusterName)
+		Expect(err).NotTo(HaveOccurred())
+		location = *oc.Location
+
+		vnet, masterSubnet, err := subnet.Split(*oc.OpenShiftClusterProperties.MasterProfile.SubnetID)
+		Expect(err).NotTo(HaveOccurred())
+		_, workerSubnet, err := subnet.Split((*(*oc.OpenShiftClusterProperties.WorkerProfiles)[0].SubnetID))
+		Expect(err).NotTo(HaveOccurred())
+
+		subnetsToReconcile = append(subnetsToReconcile, masterSubnet, workerSubnet)
```
Comment (m1kola) [inline test/e2e/operator.go:356]: Hm... Are you sure about `append` here? If we add more than one test into this `Describe` block the list will keep growing with duplicates. Or am I missing something?

File: test/e2e/operator.go [hunk 29]
```diff
@@ -321,3 +325,91 @@ var _ = XDescribe("ARO Operator - MachineSet Controller", func() {
...
+		Expect(err).NotTo(HaveOccurred())
+		resourceGroup = r.ResourceGroup
+		vnetName = r.ResourceName
+	}
+
+	// Creates an empty NSG that gets assigned to master/worker subnets.
+	createDummyNSG := func() {
+		dummynsg = mgmtnetwork.SecurityGroup{
+			Location:                      to.StringPtr(location),
+			Name:                          to.StringPtr("dummynsg"),
```
Comment (m1kola) [inline test/e2e/operator.go:368]: Is it worth adding a constant for `"dummynsg"`?

**Review by m1kola — COMMENTED** [review]
Looks good in geenral. I'm concerned a bit about `append` (see below). Rest of the comments are minor issues/suggestions.

File: test/e2e/operator.go [hunk 30]
```diff
@@ -321,3 +325,91 @@ var _ = XDescribe("ARO Operator - MachineSet Controller", func() {
...
+					return false, err
+				}
+				if *s.NetworkSecurityGroup.ID == *correctNSG {
+					log.Infof("%s subnet's nsg matched expected value", subnet)
+					return true, nil
+				}
+				log.Errorf("%s nsg: %s did not match expected value: %s", subnet, *s.NetworkSecurityGroup.ID, *correctNSG)
+				return false, nil
+			})
+			Expect(err).NotTo(HaveOccurred())
```
[the remark this answers is not in the document]
Comment (bryanro92) [reply test/e2e/operator.go:412]: I will work on these suggestions now that I have finally found the bug in the nsg /subnets reconciler https://github.com/Azure/ARO-RP/pull/1844

[the remark this answers is not in the document]
Comment (bryanro92) [reply test/e2e/operator.go]: Thank you again for your review, Mikalai. I should now be ready for another round of review with my latest changes.

Regarding the append concern. I shifted to a map [string]*string for two reasons: 1 to achieve uniqueness, 2 to populate each subnet with the value of the correct NSG which is used later in second loop that validates NSG reconciliation.

**Review by petrkotas — COMMENTED** [review]
At this point I believe we are good to go. I have one simple question and after that I am happy.

File: test/e2e/operator.go [hunk 31]
```diff
@@ -321,3 +325,98 @@ var _ = XDescribe("ARO Operator - MachineSet Controller", func() {
...
+			SecurityGroupPropertiesFormat: &mgmtnetwork.SecurityGroupPropertiesFormat{},
+		}
+		err := clients.NetworkSecurityGroups.CreateOrUpdateAndWait(ctx, resourceGroup, nsg, testnsg)
+		Expect(err).NotTo(HaveOccurred())
+		testnsg, err = clients.NetworkSecurityGroups.Get(ctx, resourceGroup, nsg, "")
+		Expect(err).NotTo(HaveOccurred())
+	}
+
+	BeforeEach(func() {
+		enableReconcileSubnet()
```
[the remark this answers is not in the document]
Comment (bryanro92) [reply test/e2e/operator.go:384]: No need to disable reconciler that I can think of. I think it's a good idea to have it on, we should be defaulting it to on shortly and we can remove that code anyways.

File: test/e2e/operator.go [hunk 32]
```diff
@@ -321,3 +325,98 @@ var _ = XDescribe("ARO Operator - MachineSet Controller", func() {
 		}
 	})
 })
+
+var _ = Describe("ARO Operator - Azure Subnet Reconciler", func() {
+	var vnetName, location, resourceGroup string
+	var subnetsToReconcile map[string]*string
```
Comment (m1kola) [inline test/e2e/operator.go:331]: I think we do not need a pointer to a string as a value. Just `string` should be ok.

```suggestion
	var subnetsToReconcile map[string]string
```

**Review by m1kola — APPROVED** [review]
Looks good to me. One minor comment, but nothing to block the PR further. Let's merge!
